### PR TITLE
[Example] add unetformer

### PR DESCRIPTION
**Pull request by tongjiqkx**
使用paddle框架复现了unetformer算法

### CODE REVIEW

**Comment by HydrogenSulfate** [comment]
@tongjiqkx 请问这个PR是PaddleScience的什么开源赛题题目？

**Comment by tongjiqkx** [comment]
你好，是这个



---- 回复的原邮件 ----
| 发件人 | ***@***.***> |
| 日期 | 2025年06月30日 09:45 |
| 收件人 | ***@***.***> |
| 抄送至 | ***@***.***>***@***.***> |
| 主题 | Re: [PaddlePaddle/PaddleScience] “unetformer” (PR #1176) |
HydrogenSulfate left a comment (PaddlePaddle/PaddleScience#1176)

@tongjiqkx 请问这个PR是PaddleScience的什么开源赛题题目？

**Comment by HydrogenSulfate** [comment]
> 你好，是这个
> […](#)
> ---- 回复的原邮件 ---- | 发件人 | ***@***.***> | | 日期 | 2025年06月30日 09:45 | | 收件人 | ***@***.***> | | 抄送至 | ***@***.***>***@***.***> | | 主题 | Re: [PaddlePaddle/PaddleScience] “unetformer” (PR #1176) | HydrogenSulfate left a comment ([PaddlePaddle/PaddleScience#1176](https://github.com/PaddlePaddle/PaddleScience/pull/1176)) @tongjiqkx 请问这个PR是PaddleScience的什么开源赛题题目？ — Reply to this email directly, view it on GitHub, or unsubscribe. You are receiving this because you were mentioned.Message ID: ***@***.***>


你的回复里没有给出有效的地址

**Comment by tongjiqkx** [comment]
卢自波老师那说是揭榜挂帅的项目



---- 回复的原邮件 ----
| 发件人 | ***@***.***> |
| 日期 | 2025年06月30日 10:14 |
| 收件人 | ***@***.***> |
| 抄送至 | ***@***.***>***@***.***> |
| 主题 | Re: [PaddlePaddle/PaddleScience] “unetformer” (PR #1176) |
HydrogenSulfate left a comment (PaddlePaddle/PaddleScience#1176)

你好，是这个
…
---- 回复的原邮件 ---- | 发件人 | @.> | | 日期 | 2025年06月30日 09:45 | | 收件人 | @.> | | 抄送至 | @.>@.> | | 主题 | Re: [PaddlePaddle/PaddleScience] “unetformer” (PR #1176) | HydrogenSulfate left a comment (PaddlePaddle/PaddleScience#1176) @tongjiqkx 请问这个PR是PaddleScience的什么开源赛题题目？ — Reply to this email directly, view it on GitHub, or unsubscribe. You are receiving this because you were mentioned.Message ID: @.***>

你的回复里没有给出有效的地址

**Comment by HydrogenSulfate** [comment]
你好，这个PR跟题目【多源遥感图像的土地覆盖场景分析】有什么具体关系吗？还是只是说复现了unetformer这个模型而已？

**Comment by tongjiqkx** [comment]
复现了unetformer具体的模型，根据别人的论文复现的



---- 回复的原邮件 ----
| 发件人 | ***@***.***> |
| 日期 | 2025年06月30日 10:27 |
| 收件人 | ***@***.***> |
| 抄送至 | ***@***.***>***@***.***> |
| 主题 | Re: [PaddlePaddle/PaddleScience] “unetformer” (PR #1176) |
HydrogenSulfate left a comment (PaddlePaddle/PaddleScience#1176)

你好，这个PR跟题目【多源遥感图像的土地覆盖场景分析】有什么具体关系吗？还是只是说复现了unetformer这个模型而已？

File: mkdocs.yml [hunk 1]
```diff
@@ -110,6 +110,7 @@ nav:
         - FengWu: zh/examples/fengwu.md
         - FuXi: zh/examples/fuxi.md
         - WGAN_GP: zh/examples/wgan_gp.md
+        - UNetFormer: zh/examples/UNetFormer.md
```
Comment (liaoxin2) [inline mkdocs.yml]: zh/examples/unetformer.md

File: docs/zh/examples/unetformer.md [hunk 2]
```diff
@@ -0,0 +1,222 @@
+# UNetFormer
+
+!!! note
+
+    1. 运行之前，建议快速了解一下[数据集](#31)和[数据读取方式](#32-dataset-api)。
+    2. 将[Vaihingen数据集]、[Potsdam数据集]、[UAVid数据集]和[LoveDA]数据集下载到`data`目录中对应的子目录（如`data/vaihingen/train_images`）。
```
Comment (liaoxin2) [inline docs/zh/examples/unetformer.md]: 只用了Vaihingen数据集，那就把其他的数据集去掉吧

File: docs/zh/examples/unetformer.md [hunk 3]
```diff
@@ -0,0 +1,222 @@
...
+\mathcal{L}_{c e}=-\frac{1}{N} \sum_{n=1}^{N} \sum_{{ }^{\prime} k=1}^{K} y_{k}^{(n)} \log \hat{y}_{k}^{(n)}
+$$  
+$$
+\mathcal{L}_{\text {dice }}=1-\frac{2}{N} \sum_{n=1}^{N} \sum_{k=1}^{K} \frac{\hat{y}_{k}^{(n)} y_{k}^{(n)}}{\hat{y}_{k}^{(n)}+y_{k}^{(n)}}
+$$  
+$$
+\mathcal{L}_{\text {p }}=\mathcal{L}_{c e}+\mathcal{L}_{\text {dice }}
+$$  
+
+其中N、K分别表示样本数量和类别数量。$y^{(n)}$和$\hat{y}^{(n)}$表示标签的one-hot编码和相应的softmax输出，n∈[1，⋯，n]。
```
Comment (liaoxin2) [inline docs/zh/examples/unetformer.md]: n∈[1，⋯，n] 需要latex语法

File: docs/zh/examples/unetformer.md [hunk 4]
```diff
@@ -0,0 +1,222 @@
...
+
+![UNetFormer1](https://i-blog.csdnimg.cn/direct/1089c5f687184ea384099b2c35bcec92.png)
+
+- `ResBlock`是resnet18网络的各个模块。
+
+- `GLTB`由全局-局部注意、MLP、两个batchnorm层和两个加和操作组成。
+
+### 2.2 损失函数
+
+判别器的损失函数由两部分组成，主损失函数$\mathcal{L}_{\text {p }}$为ce损失函数和dice损失函数。其表达式为：
```
Comment (liaoxin2) [inline docs/zh/examples/unetformer.md]: ce损失函数和dice损失函数是否是下文提到的SoftCrossEntropyLoss交叉熵损失函数和DiceLoss损失函数，两者最好保持一致

File: docs/zh/examples/unetformer.md [hunk 5]
```diff
@@ -0,0 +1,222 @@
...
+### 3.4 loss函数
+
+UNetFormer的损失函数由SoftCrossEntropyLoss交叉熵损失函数和DiceLoss损失函数组成
+
+#### 3.4.1 SoftCrossEntropyLoss
+
+
+``` py linenums="10"
+--8<--
+eexamples/UNetFormer/geoseg/losses/soft_ce.py:10:40
```
Comment (liaoxin2) [inline docs/zh/examples/unetformer.md]: eexample多了一个e

File: docs/zh/examples/unetformer.md [hunk 6]
```diff
@@ -0,0 +1,222 @@
...
+--8<--
+examples/UNetFormer/config/vaihingen/unetformer.py:62:69
+--8<--
+```
+
+### 3.6 模型训练
+
+``` py linenums="33"
+--8<--
+examples/UNetFormer/train_supervision.py:33:290
```
Comment (liaoxin2) [inline docs/zh/examples/unetformer.md]: 这个代码块不太对应

File: docs/zh/examples/unetformer.md [hunk 7]
```diff
@@ -0,0 +1,222 @@
...
+
+|  F1  |  mIOU  |  OA  |
+| :----: | :----: | :----: |
+| 0.9048 | 0.8298 | 0.9282 |
+
+
+## 6. 参考文献
+
+- [UNetFormer: A UNet-like Transformer for Efficient Semantic Segmentation of Remote Sensing Urban Scene Imagery](https://arxiv.org/abs/2109.08937)
+- [参考代码](https://github.com/WangLibo1995/GeoSeg)
```
Comment (liaoxin2) [inline docs/zh/examples/unetformer.md]: `[https://github.com/WangLibo1995/GeoSeg](https://github.com/WangLibo1995/GeoSeg)`

File: docs/zh/examples/unetformer.md [hunk 8]
```diff
@@ -0,0 +1,222 @@
...
+
+    ```
+    python GeoSeg/train_supervision.py -c GeoSeg/config/vaihingen/unetformer.py
+    ```
+
+模型评估命令  
+
+    ``` 
+    python GeoSeg/vaihingen_test.py -c GeoSeg/config/vaihingen/unetformer.py -o fig_results/vaihingen/unetformer --rgb
+    ```
```
Comment (liaoxin2) [inline docs/zh/examples/unetformer.md]: 这部分运行命令，请参考其他文档的书写格式修改一下

File: docs/zh/examples/unetformer.md [hunk 9]
```diff
@@ -0,0 +1,222 @@
...
+    --mask-dir "data/vaihingen/test_masks" \
+    --output-img-dir "data/vaihingen/test/images_1024" \
+    --output-mask-dir "data/vaihingen/test/masks_1024_rgb" \
+    --mode "val" --split-size 1024 --stride 1024 \
+    --gt
+    ```  
+模型训练命令  
+
+    ```
+    python GeoSeg/train_supervision.py -c GeoSeg/config/vaihingen/unetformer.py
```
Comment (liaoxin2) [inline docs/zh/examples/unetformer.md]: python train_supervision.py -c config/vaihingen/unetformer.py

File: docs/zh/examples/unetformer.md [hunk 10]
```diff
@@ -0,0 +1,222 @@
...
+    ```
+
+模型评估命令  
+
+    ``` 
+    python GeoSeg/vaihingen_test.py -c GeoSeg/config/vaihingen/unetformer.py -o fig_results/vaihingen/unetformer --rgb
```
Comment (liaoxin2) [inline docs/zh/examples/unetformer.md]: python vaihingen_test.py -c config/vaihingen/unetformer.py -o fig_results/vaihingen/unetformer --rgb 最好将模型文件的路径也放在命令中

File: docs/zh/examples/unetformer.md [hunk 11]
```diff
@@ -0,0 +1,222 @@
...
+
+本段落仅简单介绍模型原理，具体细节请阅读[UNetFormer: A UNet-like Transformer for Efficient 
+Semantic Segmentation of Remote Sensing Urban 
+Scene Imagery](https://arxiv.org/abs/2109.08937)。
+
+### 2.1 模型结构
+
+UNetFormer是一种基于transformer的解码器的深度学习网络，下图显示了模型的整体结构。
+
+![UNetFormer1](https://i-blog.csdnimg.cn/direct/1089c5f687184ea384099b2c35bcec92.png)
```
Comment (liaoxin2) [inline docs/zh/examples/unetformer.md]: 图片链接换为https://paddle-org.bj.bcebos.com/paddlescience/docs/unetformer/unetformer.png

File: docs/zh/examples/unetformer.md [hunk 12]
```diff
@@ -0,0 +1,222 @@
...
+--8<--
+```
+
+## 4. 结果展示
+
+使用[Vaihingen](https://www.isprs.org/resources/datasets/benchmarks/UrbanSemLab/2d-sem-label-vaihingen.aspx)数据集的训练结果。
+
+|  F1  |  mIOU  |  OA  |
+| :----: | :----: | :----: |
+| 0.9048 | 0.8298 | 0.9282 |
```
Comment (liaoxin2) [inline docs/zh/examples/unetformer.md]: 按照最新的那个模型权重文件测试的结果吧，这里修改一下

File: examples/UNetFormer/LICENSE
Comment (liaoxin2) [inline examples/UNetFormer/LICENSE:1]: 这个文件删掉

File: examples/UNetFormer/__init__.py
Comment (liaoxin2) [inline examples/UNetFormer/__init__.py:1]: 这个文件删掉

File: examples/UNetFormer/config/vaihingen/unetformer.py [hunk 13]
```diff
@@ -0,0 +1,65 @@
+import paddle
+from geoseg.datasets.vaihingen_dataset import CLASSES, VaihingenDataset, train_aug, val_aug
+from geoseg.losses import UnetFormerLoss
```
Comment (liaoxin2) [inline examples/UNetFormer/config/vaihingen/unetformer.py]: from geoseg.losses.useful_loss import UnetFormerLoss

File: docs/zh/examples/unetformer.md [hunk 14]
```diff
@@ -0,0 +1,222 @@
...
+
+$$
+\mathcal{L}_{c e}=-\frac{1}{N} \sum_{n=1}^{N} \sum_{{ }^{\prime} k=1}^{K} y_{k}^{(n)} \log \hat{y}_{k}^{(n)}
+$$  
+$$
+\mathcal{L}_{\text {dice }}=1-\frac{2}{N} \sum_{n=1}^{N} \sum_{k=1}^{K} \frac{\hat{y}_{k}^{(n)} y_{k}^{(n)}}{\hat{y}_{k}^{(n)}+y_{k}^{(n)}}
+$$  
+$$
+\mathcal{L}_{\text {p }}=\mathcal{L}_{c e}+\mathcal{L}_{\text {dice }}
+$$
```
Comment (liaoxin2) [inline docs/zh/examples/unetformer.md]: 不同行的公式之间加一个空行，第一个公式是不是多了一个{ }^{\prime}

File: docs/zh/examples/unetformer.md [hunk 15]
```diff
@@ -0,0 +1,226 @@
...
+
+## 4. 结果展示
+
+使用[Vaihingen](https://www.isprs.org/resources/datasets/benchmarks/UrbanSemLab/2d-sem-label-vaihingen.aspx)数据集的训练结果。
+
+|  F1  |  mIOU  |  OA  |
+| :----: | :----: | :----: |
+| 0.9062 | 0.8318 | 0.9283 |
+![image-vaihingen1](https://paddle-org.bj.bcebos.com/paddlescience/docs/unetformer/top_mosaic_09cm_area38_0_6.tif)
+![image-vaihingen2](https://paddle-org.bj.bcebos.com/paddlescience/docs/unetformer/result.png)
```
Comment (liaoxin2) [inline docs/zh/examples/unetformer.md]: 与前面的指标之间加一个空行，另外可以用文字适当解释一下可视化的图片

File: docs/zh/examples/unetformer.md [hunk 16]
```diff
@@ -0,0 +1,222 @@
...
+
+=== "模型评估命令"
+
+    ``` sh
+    # 下载处理好的[Vaihingen测试数据集](https://paddle-org.bj.bcebos.com/paddlescience/datasets/unetformer/test.zip)，并解压。  
+    wget -nc https://paddle-org.bj.bcebos.com/paddlescience/datasets/unetformer/test.zip -P ./data/vaihingen/
+    unzip -q ./data/vaihingen/test.zip -d data/vaihingen/
+    # 下载预训练模型文件
+    wget -nc https://paddle-org.bj.bcebos.com/paddlescience/models/unetformer/unetformer-r18-512-crop-ms-e105_epoch0_best.pdparams -P ./model_weights/vaihingen/unetformer-r18-512-crop-ms-e105/
+    python vaihingen_test.py -c config/vaihingen/unetformer.py -o fig_results/vaihingen/unetformer --rgb
```
Comment (liaoxin2) [inline docs/zh/examples/unetformer.md:49]: 这一行后面少了```

File: docs/zh/examples/unetformer.md [hunk 17]
```diff
@@ -0,0 +1,222 @@
...
+UNetFormer的损失函数由SoftCrossEntropyLoss交叉熵损失函数和DiceLoss损失函数组成
+
+#### 3.4.1 SoftCrossEntropyLoss
+
+
+``` py linenums="9"
+--8<--
+examples/unetformer/geoseg/losses/soft_ce.py:9:39
+--8<--
+```
```
Comment (liaoxin2) [inline docs/zh/examples/unetformer.md]: 代码块行数不是很对应，再检查一下吧

File: docs/zh/examples/unetformer.md [hunk 18]
```diff
@@ -0,0 +1,222 @@
...
+=== "模型训练命令"
+
+    ``` sh
+    # 将[Vaihingen数据集]下载到`data`目录中对应的子目录（如`data/vaihingen/train_images`）
+    # 创建训练数据集
+    python unetformer/tools/vaihingen_patch_split.py --img-dir "data/vaihingen/train_images" --mask-dir "data/vaihingen/train_masks" --output-img-dir "data/vaihingen/train/images_1024" --output-mask-dir "data/vaihingen/train/masks_1024" --mode "train" --split-size 1024 --stride 512
+    # 创建测试数据集
+    python unetformer/tools/vaihingen_patch_split.py --img-dir "data/vaihingen/test_images" --mask-dir "data/vaihingen/test_masks_eroded" --output-img-dir "data/vaihingen/test/images_1024" --output-mask-dir "data/vaihingen/test/masks_1024" --mode "val" --split-size 1024 --stride 1024 --eroded
+    # 创建masks_1024_rgb可视化数据集
+    python unetformer/tools/vaihingen_patch_split.py --img-dir "data/vaihingen/test_images" --mask-dir "data/vaihingen/test_masks" --output-img-dir "data/vaihingen/test/images_1024" --output-mask-dir "data/vaihingen/test/masks_1024_rgb" --mode "val" --split-size 1024 --stride 1024 --gt
```
Comment (liaoxin2) [inline docs/zh/examples/unetformer.md]: 去掉unetformer ，直接python tools/vaihingen_patch_split.py

File: .pre-commit-config.yaml [hunk 19]
```diff
@@ -1,56 +1,56 @@
...
-      - id: clang-format
-        name: clang-format
-        description: Format files with ClangFormat
-        entry: bash .clang_format.hook -i
-        language: system
-        files: \.(c|cc|cxx|cpp|cu|h|hpp|hxx|cuh|proto)$
-
-exclude: |
-  ^jointContribution/
+repos:
```
Comment (HydrogenSulfate) [inline .pre-commit-config.yaml]: 这个文件请不要改动

File: examples/unetformer/geoseg/losses/cel1.py [hunk 20]
```diff
@@ -0,0 +1,200 @@
...
+
+import paddle
+import paddle.nn.functional as F
+from paddle_utils import add_tensor_methods
+
+BINARY_MODE: str = "binary"
+MULTICLASS_MODE: str = "multiclass"
+MULTILABEL_MODE: str = "multilabel"
+EPS = 1e-10
+logger = logging.getLogger(__name__)
```
Comment (HydrogenSulfate) [inline examples/unetformer/geoseg/losses/cel1.py:12]: 建议使用: `from ppsci.utils import logger`
